### PR TITLE
Make binary_cross_entropy_with_logits composite compliant

### PR DESCRIPTION
**Pull request by drisspg**
Fixes one of the tests related to this issue #75680
Removes the skip for the composite compliance test_backward and updates implementation of binary_cross_entropy_with_logits so that the correct inplace and out of place operators are used depending on tensor subclasses being present

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -314,53 +315,101 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
...
-    if (pos_weight.defined()) {
-        // pos_weight need to be broadcasted, thus mul(target) is not inplace.
-        auto log_weight = (pos_weight - 1).mul(target).add_(1);
-        loss = (1 - target).mul_(input).add_(log_weight.mul_(((-max_val).exp_().add_((-input - max_val).exp_())).log_().add_(max_val)));
+  const Tensor& pos_weight =
+      c10::value_or_else(pos_weight_opt, [] { return Tensor(); });
+
+  Tensor loss;
+  auto max_val = (-input).clamp_min_(0);
+  auto hasSubclassTensors = at::areAnyTensorSubclassLike({target, input});
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:333]: Shouldn't this also check if weight and pos_weight are subclass?

File: aten/src/ATen/native/Loss.cpp [hunk 2]
```diff
@@ -314,53 +315,101 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
...
+  if (pos_weight.defined()) {
+    // pos_weight need to be broadcasted, thus mul(target) is not inplace.
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:335]: nit: "might need to be"

File: aten/src/ATen/native/Loss.cpp [hunk 3]
```diff
@@ -314,53 +315,101 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
...
+              .add(max_val)
+              .add((-max_val).exp().add((-input - max_val).exp()).log())
+        : (1 - target)
+              .mul_(input)
+              .add_(max_val)
+              .add_((-max_val).exp_().add_((-input - max_val).exp_()).log_());
+  }
+  if (weight.defined()) {
+    if (at::areAnyTensorSubclassLike({loss, weight})) {
+      loss.mul(weight);
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:363]: This is out of place now, you need to do `loss = `

File: aten/src/ATen/native/Loss.cpp [hunk 4]
```diff
@@ -314,53 +315,101 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
...
+    grad_input = hasSubclassTensors
+        ? t.add(1).sub(target).mul(input.sigmoid()).sub(t).mul(grad)
+        : t.add(1).sub_(target).mul_(input.sigmoid()).sub_(t).mul_(grad);
+  } else {
+    grad_input = hasSubclassTensors ? (input.sigmoid() - target).mul(grad)
+                                    : (input.sigmoid() - target).mul_(grad);
+  }
+  if (weight.defined()) {
+    if (at::areAnyTensorSubclassLike({grad_input, weight})) {
+      grad_input.mul(weight);
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:402]: grad_input =

**Review by albanD — COMMENTED** [review]
It is a bit worrying that the test don't catch the fact that weights are ignored here :(

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (drisspg) [reply aten/src/ATen/native/Loss.cpp:333]: I thought the pos_weight doesn't need to be checked because: `auto log_weight = (pos_weight - 1).mul(target).add_(1);` is the only place where it could be mutated and the first mul is out of place so it wouldn't matter.  I check whether the weight tensor is subclassed only if it is defined.. I thought that was enough ?

**Comment by ngimel** [comment]
`binary_cross_entropy_with_logits` is not a composit implicit autograd operation, so why is it failing the test? Also, it doesn't look like it's calling `resize_`, all it does is calling some inplace ops on an intermediate tensor - is it not allowed for composite ops?

**Comment by albanD** [comment]
> binary_cross_entropy_with_logits is not a composit implicit autograd operation, so why is it failing the test? 

That is a very good question! cc @zou3519 ?

> all it does is calling some inplace ops on an intermediate tensor 

It depends on which ones. You can only do inplace with Tensors that are guaranteed to be of the same "type".
For example `(input + 1).mul_(target)` would fail if input is a plain Tensor (and thus `input + 1` will also be a plain Tensor) while target is a subclass.

**Comment by drisspg** [comment]
Looks like alot of changes related to op info got merged into master and I rebased on to that branch. Honestly not even sure if the changes I made are relevant since I don't 

> 

Should I change forward formula back to what it was with mostly inplace operations that aren't switched on  subclass tensors - and keep the backward update?

**Comment by drisspg** [comment]
> > > binary_cross_entropy_with_logits is not a composit implicit autograd operation, so why is it failing the test?
> > 
> > 
> > That is a very good question! cc @zou3519 ?
> 
> Composite compliance applies to CompositeImplicitAutograd pytorch operations OR a function (in python or c++) that calls PyTorch operations (a "composite").
> 
> binary_cross_entropy_with_logits's backward formula is not a PyTorch operator -- it's a C++ function that calls Pytorch operations. The backward formula is the thing that is not composite compliant.

Should I change forward formula back to what it was with mostly inplace operations that aren't switched on  subclass tensors - and keep the backward update?

**Comment by ngimel** [comment]
@zou3519 So the fixes need to exist only in `binary_cross_entropy_with_logits_backward`, not `binary_cross_entropy_with_logits`, at least for the purposes of the test? I understand Alban's concern that even for explicit autograd op subclassing won't be propagated with inplace ops. 
Should the note about inplace ops and sublassing be added somewhere? Because currently README only warns about resizing in CompositeImplicitAutograd ops.

**Comment by drisspg** [comment]
> It is a bit worrying that the test don't catch the fact that weights are ignored here :( 
> 
> 

This got me thinking and found I wrote a terrible sampling function which I think I corrected allowing me to remove a few more skips

File: aten/src/ATen/native/Loss.cpp [hunk 5]
```diff
@@ -314,53 +315,101 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
     return grad_input;
 }
 
-Tensor binary_cross_entropy_with_logits(const Tensor& input, const Tensor& target, const c10::optional<Tensor>& weight_opt, const c10::optional<Tensor>& pos_weight_opt, int64_t reduction) {
+Tensor binary_cross_entropy_with_logits(
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:318]: Why does this function needs to change?
From the comment above, I though that only the backward needed changes?

**Comment by zou3519** [comment]
> @zou3519 So the fixes need to exist only in binary_cross_entropy_with_logits_backward, not binary_cross_entropy_with_logits, at least for the purposes of the test? I understand Alban's concern that even for explicit autograd op subclassing won't be propagated with inplace ops.

Yes, sorry for the delayed reply.

> Should the note about inplace ops and sublassing be added somewhere? Because currently README only warns about resizing in CompositeImplicitAutograd ops.

Yes... I thought it was there but apparently it's not

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/Loss.cpp]: The code as written should be sufficient

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -7811,16 +7811,24 @@ def gen_shape_kwargs():
...
+    make = partial(make_tensor, device=device, dtype=dtype)
+    make_prob = partial(make, low=0, high=1)
+
+    reductions = ("mean", "sum", "none")
+
+    shapes_and_kwargs = [
+        *[(shape, None) for shape in ((), (1,), (S,), (S, S), (S, S, S))],
+        *[((S, S), dict(reduction=reduction)) for reduction in reductions],
+        *[((S, S), dict(reduction=reduction, weight=make((S, S)))) for reduction in reductions],
+        *[((S, S), dict(reduction=reduction, pos_weight=make((S,), low=0))) for reduction in reductions]
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:7823]: I'm not very familiar with binary_cross_entropy_with_logits. Are weight and pos_weight mutually exclusive?

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -7811,16 +7811,24 @@ def gen_shape_kwargs():
...
-            yield SampleInput(input, args=(target,), kwargs=base_dict)
+    make = partial(make_tensor, device=device, dtype=dtype)
+    make_prob = partial(make, low=0, high=1)
+
+    reductions = ("mean", "sum", "none")
+
+    shapes_and_kwargs = [
+        *[(shape, None) for shape in ((), (1,), (S,), (S, S), (S, S, S))],
+        *[((S, S), dict(reduction=reduction)) for reduction in reductions],
+        *[((S, S), dict(reduction=reduction, weight=make((S, S)))) for reduction in reductions],
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:7822]: Is weight required to have the same shape as the input? Can it be different?

The docs say weight is "a manual rescaling weight if provided it’s repeated to match input tensor shape" [link](https://pytorch.org/docs/stable/generated/torch.nn.functional.binary_cross_entropy_with_logits.html?highlight=binary_cross_entropy_with_logits#torch.nn.functional.binary_cross_entropy_with_logits) which implies that it can be different, but I'm not sure

**Review by zou3519 — COMMENTED** [review]
Fix for composite compliance LGTM. Since we're modifying the OpInfo, I have some questions about the semantics of the operator

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (drisspg) [reply torch/testing/_internal/common_methods_invocations.py:7823]: Ohh good catch they don't need be mutually exclusive, will update

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (drisspg) [reply torch/testing/_internal/common_methods_invocations.py:7822]: From playing around it appears that it just has to be broadcast-able with the input shape. Should I add more tests with varying broadcast-able shapes?

**Comment by drisspg** [comment]
Just Rebased onto master and this skip:
```
DecorateInfo(
                unittest.expectedFailure,
                'TestCompositeCompliance',
                'test_forward_ad'),
```
Is now failing for performing inplace ops on tensor subclass. I am going to add back in this skip and not update the forward pass because see above convseration

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py]: It would be good to add more tests with a broadcast-able shape.

Comment (drisspg) [reply torch/testing/_internal/common_methods_invocations.py]: Makes sense, just added

**Review by zou3519 — APPROVED** [review]
Thank you!
